### PR TITLE
add new parser - TFSec

### PR DESCRIPTION
**Pull request by ansidorov**
tfsec uses static analysis of your terraform templates to spot potential security issues.

It could be found at: https://github.com/tfsec/tfsec

### CODE REVIEW

**Review by valentijnscholten — APPROVED** [review]
Can you check if we need to set dedupe algorithm to HASH_CODE in settings? And if we need a hash code field config or can use the default? (Might be good to check the same for the previous parsers by you which we already merged)

**Comment by ansidorov** [comment]
> Can you check if we need to set dedupe algorithm to HASH_CODE in settings? And if we need a hash code field config or can use the default? (Might be good to check the same for the previous parsers by you which we already merged)

If I understand correctly, we need to use the default for tfsec, kics and dockle parsers.

**Comment by ansidorov** [comment]
> Can you check if we need to set dedupe algorithm to HASH_CODE in settings? And if we need a hash code field config or can use the default? (Might be good to check the same for the previous parsers by you which we already merged)

I thought again and adjusted the deduplication algorithm.

File: dojo/tools/tfsec/parser.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        return "TFSec Scan"
+
+    def get_description_for_scan_types(self, scan_type):
+        return "Import JSON output for TFSec scan report."
+
+    def get_findings(self, filename, test):
+        data = json.load(filename)
+        dupes = {}
+        if data.get('results'):
+            for item in data['results']:
```
Comment (damiencarol) [inline dojo/tools/tfsec/parser.py:31]:
```suggestion
        if 'results' not in data:
            raise ValueError("Incorrect TFSec scan, missing attribute 'results'")
        for item in data.get('results', []):
```
This is the safest way to do it

File: dojo/tools/tfsec/parser.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                        title=f"{rule_description}",
+                        test=test,
+                        severity=severity,
+                        description=description,
+                        active=active,
+                        mitigation=resolution,
+                        impact=impact,
+                        file_path=file,
+                        line=start_line,
+                        component_name=rule_provider,
```
Comment (damiencarol) [inline dojo/tools/tfsec/parser.py:68]:
```suggestion
```
`component_name` is dedicated to third party component, not sure the rule provider is the best data here.

File: dojo/unittests/tools/test_tfsec_parser.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            finding = findings[0]
+            self.assertEqual("Potentially sensitive data stored in block attribute.", finding.title)
+            self.assertEqual("Medium", finding.severity)
+            self.assertIsNotNone(finding.description)
+            self.assertTrue(finding.active)
+            self.assertEqual("Don't include sensitive data in blocks", finding.mitigation)
+            self.assertEqual("Block attribute could be leaking secrets", finding.impact)
+            self.assertEqual("tfsec-test/identity.tf", finding.file_path)
+            self.assertEqual(226, finding.line)
+            self.assertEqual("general", finding.component_name)
```
Comment (damiencarol) [inline dojo/unittests/tools/test_tfsec_parser.py:30]:
```suggestion
```
See previous comment

File: dojo/unittests/tools/test_tfsec_parser.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            finding = findings[0]
+            self.assertEqual("Legacy client authentication methods utilized.", finding.title)
+            self.assertEqual("High", finding.severity)
+            self.assertIsNotNone(finding.description)
+            self.assertTrue(finding.active)
+            self.assertEqual("Use service account or OAuth for authentication", finding.mitigation)
+            self.assertEqual("Username and password authentication methods are less secure", finding.impact)
+            self.assertEqual("tfsec-test/cluster.tf", finding.file_path)
+            self.assertEqual(52, finding.line)
+            self.assertEqual("google", finding.component_name)
```
Comment (damiencarol) [inline dojo/unittests/tools/test_tfsec_parser.py:50]:
```suggestion
```
See previous comment

File: dojo/unittests/tools/test_tfsec_parser.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            finding = findings[1]
+            self.assertEqual("Pod security policy enforcement not defined.", finding.title)
+            self.assertEqual("High", finding.severity)
+            self.assertIsNotNone(finding.description)
+            self.assertTrue(finding.active)
+            self.assertEqual("Use security policies for pods to restrict permissions to those needed to be effective", finding.mitigation)
+            self.assertEqual("Pods could be operating with more permissions than required to be effective", finding.impact)
+            self.assertEqual("tfsec-test/cluster.tf", finding.file_path)
+            self.assertEqual(52, finding.line)
+            self.assertEqual("google", finding.component_name)
```
Comment (damiencarol) [inline dojo/unittests/tools/test_tfsec_parser.py:64]:
```suggestion
```
See previous comment

File: dojo/unittests/tools/test_tfsec_parser.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            finding = findings[3]
+            self.assertEqual("Potentially sensitive data stored in block attribute.", finding.title)
+            self.assertEqual("Medium", finding.severity)
+            self.assertIsNotNone(finding.description)
+            self.assertTrue(finding.active)
+            self.assertEqual("Don't include sensitive data in blocks", finding.mitigation)
+            self.assertEqual("Block attribute could be leaking secrets", finding.impact)
+            self.assertEqual("tfsec-test/identity.tf", finding.file_path)
+            self.assertEqual(226, finding.line)
+            self.assertEqual("general", finding.component_name)
```
Comment (damiencarol) [inline dojo/unittests/tools/test_tfsec_parser.py:92]:
```suggestion
```
See previous comment

File: dojo/unittests/tools/test_tfsec_parser.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            finding = findings[2]
+            self.assertEqual("Shielded GKE nodes not enabled.", finding.title)
+            self.assertEqual("High", finding.severity)
+            self.assertIsNotNone(finding.description)
+            self.assertTrue(finding.active)
+            self.assertEqual("Enable node shielding", finding.mitigation)
+            self.assertEqual("Node identity and integrity can't be verified without shielded GKE nodes", finding.impact)
+            self.assertEqual("tfsec-test/cluster.tf", finding.file_path)
+            self.assertEqual(52, finding.line)
+            self.assertEqual("google", finding.component_name)
```
Comment (damiencarol) [inline dojo/unittests/tools/test_tfsec_parser.py:78]:
```suggestion
```
See previous comment

**Review by damiencarol — CHANGES_REQUESTED** [review]

File: dojo/settings/settings.dist.py [hunk 8]
```diff
@@ -812,6 +812,7 @@ def generate_url(scheme, double_slashes, user, password, host, port, path, param
     'DSOP Scan': ['cve'],
     'Acunetix Scan': ['title', 'description'],
     'Trivy Scan': ['title', 'severity', 'cve', 'cwe'],
+    'TFSec Scan': ['severity', 'vuln_id_from_tool', 'file_path', 'line', 'description'],
```
Comment (damiencarol) [inline dojo/settings/settings.dist.py:815]: Should we remove the description? I have the feeling that if `vuln_id_from_tool` is used in de-duplication algorithm we can ignore description.

**Review by damiencarol — APPROVED** [review]
Minor comment on de-duplication configuration but for me it's good.

File: dojo/tools/tfsec/parser.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+    def get_description_for_scan_types(self, scan_type):
+        return "Import JSON output for TFSec scan report."
+
+    def get_findings(self, filename, test):
+        data = json.load(filename)
+        dupes = {}
+        if 'results' not in data:
+            raise ValueError("Incorrect TFSec scan, missing attribute 'results'")
+        for item in data.get('results', []):
```
Comment (damiencarol) [inline dojo/tools/tfsec/parser.py:32]:
```suggestion
        if data.get('results') is None:
            return list()
        for item in data.get('results'):
```

File: dojo/settings/settings.dist.py (hunk 8 above)
Comment (ansidorov) [reply dojo/settings/settings.dist.py]: Done

**Pull request merged**